File: internal/mqttservice/mqtt.go
package mqttservice

import (
	"encoding/json"
	"fmt"
	"log"
	"time"

	mqtt "github.com/eclipse/paho.mqtt.golang"
)

type GPSTelemetryPayload struct {
	DriverID  string    `json:"driver_id"`
	Latitude  float64   `json:"latitude"`
	Longitude float64   `json:"longitude"`
	Timestamp time.Time `json:"timestamp"`
}

type MQTTBroker struct {
	client mqtt.Client
}

func NewMQTTBroker(brokerURL string) *MQTTBroker {
	opts := mqtt.NewClientOptions().AddBroker(brokerURL)
	opts.SetClientID("avandab_backend_server")
	opts.SetKeepAlive(60 * time.Second)
	opts.SetPingTimeout(10 * time.Second)

	client := mqtt.NewClient(opts)
	if token := client.Connect(); token.Wait() && token.Error() != nil {
		log.Printf("[MQTT WARNING] Could not connect to MQTT Broker (%s): %v (Running fallback mode)", brokerURL, token.Error())
	} else {
		log.Println("[MQTT SUCCESS] Connected to MQTT Telemetry Broker at", brokerURL)
	}

	b := &MQTTBroker{client: client}
	b.subscribeTelemetry()
	return b
}

func (b *MQTTBroker) subscribeTelemetry() {
	if !b.client.IsConnected() {
		return
	}
	topic := "avandab/telemetry/drivers/+/gps"
	b.client.Subscribe(topic, 1, func(c mqtt.Client, m mqtt.Message) {
		var payload GPSTelemetryPayload
		if err := json.Unmarshal(m.Payload(), &payload); err == nil {
			log.Printf("[MQTT TELEMETRY RECV] Driver %s -> Lat: %.4f, Lng: %.4f", payload.DriverID, payload.Latitude, payload.Longitude)
		}
	})
}

func (b *MQTTBroker) PublishTripUpdate(driverID string, tripID string, status string) {
	if !b.client.IsConnected() {
		return
	}
	topic := fmt.Sprintf("avandab/trips/drivers/%s/updates", driverID)
	data, _ := json.Marshal(map[string]string{
		"trip_id": tripID,
		"status":  status,
		"time":    time.Now().Format(time.RFC3339),
	})
	b.client.Publish(topic, 1, false, data)
}

File: mobile/src/services/mqtt.ts
import mqtt from 'mqtt';
import { getMQTTBrokerURL } from '../constants/network';
import { useAuthStore } from '../stores/authStore';

export interface TripDispatchUpdate {
  trip_id: string;
  status: string;
  time: string;
}

type DispatchListener = (update: TripDispatchUpdate) => void;

class MQTTTelemetryService {
  private client: mqtt.MqttClient | null = null;
  private isConnected = false;
  private dispatchListeners: Set<DispatchListener> = new Set();

  /**
   * Subscribe to in-app dispatch notifications (trip assignments/status
   * changes pushed over `avandab/trips/drivers/{id}/updates`). Returns an
   * unsubscribe function.
   */
  onDispatch(listener: DispatchListener): () => void {
    this.dispatchListeners.add(listener);
    return () => this.dispatchListeners.delete(listener);
  }

  private emitDispatch(update: TripDispatchUpdate): void {
    this.dispatchListeners.forEach((fn) => {
      try {
        fn(update);
      } catch {
        // listener errors never break the MQTT loop
      }
    });
  }

  connect(driverId: string): void {
    try {
      const brokerUrl = getMQTTBrokerURL();
      const token = useAuthStore.getState().token;

      const options: mqtt.IClientOptions = {
        clientId: `driver_${driverId}_${Math.random().toString(16).substring(2, 8)}`,
        keepalive: 60,
        reconnectPeriod: 5000,
        username: driverId,
      };
      if (token) {
        options.password = token;
      }

      // Connect to MQTT Broker over WebSockets
      this.client = mqtt.connect(brokerUrl, options);

      this.client.on('connect', () => {
        this.isConnected = true;
        console.log('[MQTT MOBILE SUCCESS] Connected to MQTT Telemetry Broker');

        // Subscribe to driver dispatch assignment updates
        const updateTopic = `avandab/trips/drivers/${driverId}/updates`;
        this.client?.subscribe(updateTopic, (err) => {
          if (!err) {
            console.log(`[MQTT SUBSCRIBED] Listening on topic: ${updateTopic}`);
          }
        });
      });

      this.client.on('message', (topic, message) => {
        console.log(`[MQTT RECV] Topic: ${topic} Payload: ${message.toString()}`);
        if (topic.includes('/updates')) {
          try {
            const parsed = JSON.parse(message.toString());
            if (parsed && typeof parsed.trip_id === 'string') {
              this.emitDispatch({
                trip_id: parsed.trip_id,
                status: typeof parsed.status === 'string' ? parsed.status : '',
                time: typeof parsed.time === 'string' ? parsed.time : '',
              });
            }
          } catch {
            // non-JSON payload — log only
          }
        }
      });

      this.client.on('error', (err) => {
        console.log('[MQTT MOBILE WARNING] Connection warning (fallback to HTTP):', err.message);
      });
    } catch (e: any) {
      console.log('[MQTT INIT WARNING]', e.message);
    }
  }

  // Publish high-frequency live GPS coordinates over MQTT
  publishLocation(driverId: string, latitude: number, longitude: number): void {
    if (this.client && this.isConnected) {
      const topic = `avandab/telemetry/drivers/${driverId}/gps`;
      const payload = JSON.stringify({
        driver_id: driverId,
        latitude,
        longitude,
        timestamp: new Date().toISOString(),
      });
      this.client.publish(topic, payload, { qos: 1 });
      console.log(`[MQTT PUBLISHED GPS] Lat: ${latitude.toFixed(4)}, Lng: ${longitude.toFixed(4)} -> ${topic}`);
    }
  }

  disconnect(): void {
    if (this.client) {
      this.client.end();
      this.isConnected = false;
    }
  }
}

export const MQTT = new MQTTTelemetryService();
